apps/eth_blockchain/lib/eth_blockchain/transaction.ex

```diff
@@ -137,31 +127,73 @@ defmodule EthBlockchain.Transaction do
          adapter,
          pid
        ) do
-    with {:ok, encoded_abi_data} <- ABIEncoder.transfer(to, amount),
-         {:ok, nonce} <- get_transaction_count(%{address: from}, adapter, pid) do
-      gas_limit = Application.get_env(:eth_blockchain, :default_contract_transaction_gas_limit)
-      gas_price = get_gas_price_or_default(attrs)
-
+    with {:ok, meta} <-
+           get_transaction_meta(attrs, :default_contract_transaction_gas_limit, adapter, pid),
+         {:ok, encoded_abi_data} <- ABIEncoder.transfer(to, amount) do
       %__MODULE__{
-        gas_limit: gas_limit,
-        gas_price: gas_price,
-        nonce: int_from_hex(nonce),
         to: from_hex(contract_address),
         data: encoded_abi_data
       }
+      |> Map.merge(meta)
       |> sign_and_hash(from)
       |> send_raw(adapter, pid)
     else
       error -> error
     end
   end
 
-  defp get_gas_price_or_default(%{gas_price: gas_price}), do: gas_price
+  @doc """
+  Submit a contract creation transaction with the given data
+  Returns {:ok, tx_hash, contract_address} if success,
+  {:error, code} || {:error, code, message} otherwise
+  """
+  def create_contract(%{from: from, contract_data: init} = attrs, adapter, pid) do
+    case get_transaction_meta(attrs, :default_contract_creation_gas_limit, adapter, pid) do
+      {:ok, %{nonce: nonce} = meta} ->
+        contract_address = get_contract_address(nonce, from)
+
+        %__MODULE__{init: from_hex(init)}
+        |> Map.merge(meta)
+        |> sign_and_hash(from)
+        |> send_raw(adapter, pid)
+        |> append_contract_address(contract_address)
+
+      error ->
+        error
+    end
+  end
+
+  defp get_transaction_meta(%{from: from} = attrs, gas_limit_type, adapter, pid) do
+    case get_transaction_count(%{address: from}, adapter, pid) do
+      {:ok, nonce} ->
+        gas_limit = get_gas_limit_or_default(gas_limit_type, attrs)
+        gas_price = get_gas_price_or_default(attrs)
+
+        {:ok, %{gas_price: gas_price, gas_limit: gas_limit, nonce: int_from_hex(nonce)}}
+
+      error ->
+        error
+    end
+  end
+
+  defp get_contract_address(nonce, sender) do
+    "0x" <> <<_::bytes-size(24)>> <> contract_address =
+      [from_hex(sender), encode_unsigned(nonce)]
+      |> ExRLP.encode()
+      |> Keccak.kec()
+      |> to_hex()
 
-  defp get_gas_price_or_default(_attrs) do
-    Application.get_env(:eth_blockchain, :default_gas_price)
+    "0x" <> contract_address
   end
 
+  defp get_gas_limit_or_default(_type, %{gas_limit: gas_limit}), do: gas_limit
```

Comment:
All those gas helping functions could be moved to another module as well, if we decide to extract the contract creating stuff to its own module.